src/server/transaction.cc

```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
+  // and dispatch a single task to the its shard set.
+  if ((unique_shard_cnt_ == 1) && !IsGlobal() && !multi_) {
     DCHECK_EQ(1u, shard_data_.size());
 
+    // IsArmedInShard() first checks run_count_ before shard_data, so use release ordering.
     shard_data_[0].local_mask |= ARMED;
-
-    // memory_order_release because we do not want it to be reordered with shard_data writes
-    // above.
-    // IsArmedInShard() first checks run_count_ before accessing shard_data.
     run_count_.fetch_add(1, memory_order_release);
+
     time_now_ms_ = GetCurrentTimeMs();
 
-    // Please note that schedule_cb can not update any data on ScheduleSingleHop stack when
-    // run_fast is false.
-    // since ScheduleSingleHop can finish before ScheduleUniqueShard returns.
-    // The problematic flow is as follows: ScheduleUniqueShard schedules into TxQueue
-    // (hence run_fast is false), and then calls PollExecute that in turn runs
-    // the callback which calls DecreaseRunCnt.
-    // As a result WaitForShardCallbacks below is unblocked before schedule_cb returns.
-    // However, if run_fast is true, then we may mutate stack variables, but only
-    // before DecreaseRunCnt is called.
-    auto schedule_cb = [&] {
+    // NOTE: schedule_cb cannot update data on stack when run_fast is false.
+    // This is because this function can finish before the callback returns.
+
+    // This happens when ScheduleUniqueShard schedules into TxQueue (hence run_fast is false), and
+    // then calls PollExecute that in turn runs the callback which calls DecreaseRunCnt. As a result
+    // WaitForShardCallbacks below is unblocked before schedule_cb returns. However, if run_fast is
+    // true, then we may mutate stack variables, but only before DecreaseRunCnt is called.
+    auto schedule_cb = [this, &was_ooo] {
       bool run_fast = ScheduleUniqueShard(EngineShard::tlocal());
       if (run_fast) {
-        run_eager = true;
+        was_ooo = true;
         // it's important to DecreaseRunCnt only for run_fast and after run_eager is assigned.
         // If DecreaseRunCnt were called before ScheduleUniqueShard finishes
         // then WaitForShardCallbacks below could exit before schedule_cb assigns return value
         // to run_eager and cause stack corruption.
         CHECK_GE(DecreaseRunCnt(), 1u);
       }
     };
-
     shard_set->Add(unique_shard_id_, std::move(schedule_cb));  // serves as a barrier.
   } else {
-    // Transaction spans multiple shards or it's global (like flushdb) or multi.
-    // Note that the logic here is a bit different from the public Schedule() function.
-    if (multi_) {
-      if (multi_->is_expanding)
-        multi_->AddLocks(Mode());
-    } else {
+    // This transaction either spans multiple shards and/or is multi.
+
+    if (!multi_)  // Multi schedule in advance.
       ScheduleInternal();
-    }
+
+    if (multi_ && multi_->is_expanding)
```

Comment:
nit: why did you change it? now we have 2 checks `if (!multi)`, `if (multi)` instead of a single check

Reply:
It was a nested if before, which in my mind is harder to read, so I made it a single if-else if. And instead of this nested if, I just split it up. Now we have a easily visible condition->action part